llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp

```diff
@@ -0,0 +1,643 @@
+    MIRBuilder.buildInstr(SPIRV::OpReturn);
+
+    return MBB;
+  }
+
+  // Replace switches with a single target with an unconditional branch.
+  bool replaceEmptySwitchWithBranch(MachineFunction &MF) {
+    bool Modified = false;
+    for (MachineBasicBlock &MBB : MF) {
+      MachineInstr *I = &*MBB.rbegin();
+      GIntrinsic *II = dyn_cast<GIntrinsic>(I);
+      if (!II || II->getIntrinsicID() != Intrinsic::spv_switch ||
+          II->getNumOperands() > 3)
+        continue;
+
+      Modified = true;
+      assert(II->getOperand(2).isMBB());
+      MachineBasicBlock *Target = II->getOperand(2).getMBB();
+
+      MachineIRBuilder MIRBuilder(MF);
+      MIRBuilder.setInsertPt(MBB, MBB.end());
+      MIRBuilder.buildBr(*Target);
+      MBB.erase(I);
+    }
+
+    return Modified;
+  }
+
+  // Traverse each loop, and adds an OpLoopMerge instruction to its header
+  // that respect the convergence region node it belongs to.
+  // The Continue target is the only back-edge in that loop.
+  // The merge target is the only exiting node of the convergence region.
+  bool addMergeForLoops(MachineFunction &MF) {
+    auto &TII = *MF.getSubtarget<SPIRVSubtarget>().getInstrInfo();
+    auto &TRI = *MF.getSubtarget<SPIRVSubtarget>().getRegisterInfo();
+    auto &RBI = *MF.getSubtarget<SPIRVSubtarget>().getRegBankInfo();
+
+    const auto &MLI = getAnalysis<MachineLoopInfoWrapperPass>().getLI();
+    const auto *TopLevelRegion =
+        getAnalysis<SPIRVConvergenceRegionAnalysisWrapperPass>()
+            .getRegionInfo()
+            .getTopLevelRegion();
+
+    bool Modified = false;
+    for (auto &MBB : MF) {
+      // Not a loop header. Ignoring for now.
+      if (!MLI.isLoopHeader(&MBB))
+        continue;
+      auto *L = MLI.getLoopFor(&MBB);
+
+      // This loop header is not the entrance of a convergence region. Ignoring
+      // this block.
+      auto *CR = getRegionForHeader(TopLevelRegion, &MBB);
+      if (CR == nullptr)
+        continue;
+
+      auto *Merge = getExitFor(MF, CR);
+      // This is a special case:
+      // We are indeed in a loop, but there are no exits (infinite loop).
+      // This means the actual branch is unconditional, hence won't require any
+      // OpLoopMerge.
+      if (Merge == nullptr) {
+        Merge = createReturnBlock(MF, MBB);
+      }
+
+      auto *Continue = L->getLoopLatch();
+
+      // Conditional branch are built using a fallthrough if false + BR.
+      // So the last instruction is not always the first branch.
+      auto *I = &*MBB.getFirstTerminator();
+      BuildMI(MBB, I, I->getDebugLoc(), TII.get(SPIRV::OpLoopMerge))
+          .addMBB(Merge)
+          .addMBB(Continue)
+          .addImm(SPIRV::SelectionControl::None)
+          .constrainAllUses(TII, TRI, RBI);
+      Modified = true;
+    }
+
+    return Modified;
+  }
+
+  // Add an OpSelectionMerge to each node with an out-degree of 2 or more.
+  bool addMergeForConditionalBranches(MachineFunction &MF) {
```

Comment:
How do you make sure that all blocks with in-degree of 2 or more are the merge block of a header? The example you give below is a good example?

```
//          A
//         / \
//        B<--C
//        \    \
//         \    D
//          \  /
//           E
```

From what I can tell, this will add two `OpSelectionMerge` instructions. One in `A` and one in `C`. Both will use `E` as the merge block. `B` will not be a merge block. This violates the rules for maximal reconvergence (https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_maximal_reconvergence.html).

Note, we want to use maximal reconvergence now that it is released. Otherwise, many cases would have implementation defined behaviour.